dpnp/linalg/dpnp_utils_linalg.py

```diff
@@ -96,85 +96,74 @@ def _batched_eigh(a, UPLO, eigen_mode, w_type, v_type):
     lapack_func = (
-        "_heevd" if dpnp.issubdtype(v_type, dpnp.complexfloating) else "_syevd"
+        "_heevd_batch"
+        if dpnp.issubdtype(v_type, dpnp.complexfloating)
+        else "_syevd_batch"
     )
 
     a_sycl_queue = a.sycl_queue
-    a_order = "C" if a.flags.c_contiguous else "F"
+
+    a_orig_shape = a.shape
+    a_orig_order = "C" if a.flags.c_contiguous else "F"
+    # get 3d input array by reshape
+    a = dpnp.reshape(a, (-1, a_orig_shape[-2], a_orig_shape[-1]))
 
     _manager = dpu.SequentialOrderManager[a_sycl_queue]
+    dep_evs = _manager.submitted_events
 
-    # need to loop over the 1st dimension to get eigenvalues and
-    # eigenvectors of 3d matrix A
-    batch_size = a.shape[0]
-    eig_vecs = [None] * batch_size
-    for i in range(batch_size):
-        # oneMKL LAPACK assumes fortran-like array as input, so
-        # allocate a memory with 'F' order for dpnp array of eigenvectors
-        eig_vecs[i] = dpnp.empty_like(a[i], order="F", dtype=v_type)
+    # Reorder the elements by moving the last two axes of `a` to the front
+    # to match fortran-like array order which is assumed by syevd/heevd.
+    a = dpnp.moveaxis(a, (-2, -1), (0, 1))
+    a_usm_arr = dpnp.get_usm_ndarray(a)
 
-        # use DPCTL tensor function to fill the array of eigenvectors with
-        # content of input array
-        ht_ev, copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
-            src=a_usm_arr[i],
-            dst=eig_vecs[i].get_array(),
-            sycl_queue=a_sycl_queue,
-            depends=_manager.submitted_events,
-        )
-        _manager.add_event_pair(ht_ev, copy_ev)
+    a_copy = dpnp.empty_like(a, dtype=v_type, order="F")
+    ht_ev, a_copy_ev = ti._copy_usm_ndarray_into_usm_ndarray(
+        src=a_usm_arr,
+        dst=a_copy.get_array(),
+        sycl_queue=a_sycl_queue,
+        depends=dep_evs,
+    )
+    _manager.add_event_pair(ht_ev, a_copy_ev)
 
-        # TODO: Remove this w/a when MKLD-17201 is solved.
-        # Waiting for a host task executing an OneMKL LAPACK syevd/heevd call
-        # on CPU causes deadlock due to serialization of all host tasks
-        # in the queue.
-        # We need to wait for each host tasks before calling _seyvd and _heevd
-        # to avoid deadlock.
-        if is_cpu_device:
-            dpnp.synchronize_array_data(a)
+    # allocate a memory for dpnp array of eigenvalues
+    w = dpnp.empty_like(a, shape=a_orig_shape[:-1], dtype=w_type)
+    w_orig_shape = w.shape
 
-        # call LAPACK extension function to get eigenvalues and
-        # eigenvectors of a portion of matrix A
-        ht_ev, lapack_ev = getattr(li, lapack_func)(
-            a_sycl_queue,
-            jobz,
-            uplo,
-            eig_vecs[i].get_array(),
-            w[i].get_array(),
-            depends=[copy_ev],
-        )
-        _manager.add_event_pair(ht_ev, lapack_ev)
+    # get 2d dpnp array with eigenvalues by reshape
+    w = w.reshape(-1, w_orig_shape[-1])
```

Comment:
What is the reason `w` isn't allocated with expected shape at the beginning at line#140? Why do we need to have an extra step with reshape?

Reply:
This has been changed. Thank you